patches/minecraft/net/minecraft/world/WorldEntitySpawner.java.patch

```diff
@@ -50,7 +50,21 @@
                                                      {
                                                          continue label134;
                                                      }
-@@ -228,8 +232,9 @@
+@@ -219,6 +223,13 @@
+         }
+         else
+         {
++            return net.minecraftforge.common.ForgeHooks.canSpawnAt(p_180267_0_, p_180267_1_, p_180267_2_);
++        }
++    }
++
++    public static boolean canCreatureTypeSpawnBody(EntityLiving.SpawnPlacementType p_180267_0_, World p_180267_1_, BlockPos p_180267_2_)
++    {
```

Comment:
This body needs to be moved way down, into the final else section. As Blocks *should* still be able to veto the spawning. And we want to also prevent spawning in barriers and bedrock. Basically the only like you want to replace is empty space checks?

Reply:
I do think it is good for mods to try to respect `canCreatureSpawn` in general, but there are some issues with forcing it before the hook here:

One issue with that is that, for the example case in the test mod, ice blocks would be vetoing the spawning due to the default implementation of `canCreatureSpawn`, so that example would not actually be possible to have.

Another point is that, currently, `IN_WATER` doesn't perform those checks, and the issue with requiring them here is that it prevents custom spawn logic from determining which positions around the entity should be checked - it would prevent mods from having `IN_LAVA` or `ON_CEILING` spawn placement types, for example.